wazuh/wazuh
# Enhancement `versionMatcher` testtool

## Objective
The purpose of this PR is to enhance the audit and validation process of package matches during the vulnerability detection scan. By addressing the issue outlined in Issue #21971, we aim to introduce a bash script that reads logs of package matches, extracts versions, and utilizes `rpmdev-vercmp` or `dpkg` tools for accurate version validation. This enhancement is designed to operate within a Dockerized native OS environment, with the script set as the entry point, thereby ensuring compatibility and reproducibility across different systems.

## Description
- Developed a new bash script `validate-version-matches.sh` that automates the process of reading vulnerability detection scan logs, extracting package versions, and validating them using `rpmdev-vercmp` for RPM packages and `dpkg` for DEB packages.
- The script is designed to be Docker-friendly, with Dockerfile adjustments made to build a native OS environment (both CentOS for RPM and Debian for DEB validations) where the script is copied and set as the Docker entry point. This ensures that validations are performed in a clean, controlled environment.
- Special attention was given to error handling and reporting within the script to ensure that any discrepancies or validation failures are clearly logged, providing valuable insights for auditing and further investigation.

## Quality Assurance

### Static Analysis
- Conducted static analysis on the bash script to ensure best practices in scripting, readability, and security. Tools such as ShellCheck were used to identify and correct any potential issues.

### Testing

#### Unit Testing
N/A

#### Integration Testing
- The entire process, from log reading through to version validation within the Dockerized environment, was thoroughly tested. This included creating mock vulnerability detection logs with known version matching scenarios to ensure the script accurately validates and reports.
- **Testing Steps**:
  1. Build the Docker images for both RPM and DEB environments using the provided Dockerfile.
  2. Run the Docker containers with the vulnerability detection logs as input.
  3. Review the script output for accuracy in version extraction and validation, comparing against expected results.
  4. Analyze any reported discrepancies to validate the effectiveness of the script.


Example

```text
# Content of ubuntu_match.log
2024/02/19 14:33:03 wazuh-modulesd:vulnerability-scanner[66172] packageScanner.hpp:139 at operator()(): DEBUG: Match found, the package 'mysql', is vulnerable to 'CVE-2011-2262'. Current version: '5.5.23' is equal to '5.5.23'. - Agent 'ubuntu22' (ID: '001', Version: 'v4.7.0').
2024/02/19 10:39:33 wazuh-modulesd:vulnerability-scanner[49303] packageScanner.hpp:185 at operator()(): DEBUG: Match found, the package 'filebeat', is vulnerable to 'CVE-2023-31413'. Current version: '7.10.2' (less than '' or equal to '7.17.9'). - Agent 'localhost' (ID: '000', Version: 'v4.8.0').
```

```bash
docker run --rm -v ./ReportsDEB:/usr/local/bin/version_verifier/Reports version_verifier_dpkg
=== Version Verification ===
=== Version Verification Completed ===
```

With the logs of #21911 

```bash
docker run --rm -v ./Reports:/usr/local/bin/version_verifier/Reports version_verifier_rpm
=== Version Verification ===
WARNING: hyphen in release2: 11-16
WARNING: hyphen in release2: 11-16
=== Version Verification Completed ===
```

## Additional Notes
- Future enhancements may include extending the script's capabilities to support additional package managers.
